### DISCUSSION TITLE
Typescript type annotations for tree-sitter grammars

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1334] [Q&A] [answered]
I'm a typescript beginner, but as I understand it typescript can import node modules and perform type checking as long as the imported node module has a "d.ts" file providing type information about the module's API. The tree-sitter node module seems to have this, as I can import it without issue:
```ts
import Parser = require('tree-sitter');
```
However, if I try to do the same thing with my own grammar:
```ts
import TlaPlus = require('@tlaplus/tree-sitter-tlaplus')
```
then I get an error:
>error TS7016: Could not find a declaration file for module '@tlaplus/tree-sitter-tlaplus'. 'C:/Users/ahelwer/source/tlaplus/tree-sitter-tlaplus/bindings/node/index.js' implicitly has an 'any' type.     
  Try `npm i --save-dev @types/tlaplus__tree-sitter-tlaplus` if it exists or add a new declaration (.d.ts) file containing `declare module '@tlaplus/tree-sitter-tlaplus';`

So I have to silence this with the following annotation, although then no type checking can be performed for my grammar object:
```ts
//@ts-ignore
import TlaPlus = require('@tlaplus/tree-sitter-tlaplus')
```
I guess the d.ts file isn't auto-generated along with the rest of the node bindings for a grammar. Has anyone created such a file? Is there any real benefit to doing so, or are all the interesting types declared in the tree-sitter module anyway?

**Comment by razzeee** [comment, marked as answer]
Have a look at https://www.npmjs.com/package/@asgerf/dts-tree-sitter

I think we're doing it in our grammar, mostly for the `fields`

**Reply by ahelwer** [reply]
Very neat project, thanks!
